guardian/prosemirror-elements
# Add NestedElementField

Co-authored with @davidfurey and @jonathonherbert - first few commits were David's but I unintentionally rebased his name out of the commit history.

<img width="726" alt="image" src="https://github.com/guardian/prosemirror-elements/assets/34686302/8d740834-5785-4986-b1de-5c1f3bc108bc">


## What does this change?
This PR adds the nested element field. This field is a rich text editor that is itself able to contain elements. This will enable us to build larger, 'structural' elements within Composer that can contain elements like images and videos.

There is some further work to do in future PRs:
- Add some helper functions to provide a content expression for a nested element field that disallows certain elements (for example, we probably don't want elements that contain nested element fields within other nested element fields).
- There is a known bug in repeater fields involving applying Decorations . Trying to insert an element (e.g. via pasting) into any node other than the initial one in a repeater field now works, but the Decorations for notes and Typerighter matches within nestedElements aren't currently working.
- Make sure that the element controls are usable within the field. (e.g. move, delete, select)

This PR could do with some tests for nested element fields - I'll add some when I'm back from holiday.

## How to test
1. Run the demo app for this branch with `yarn start`.
2. Add the 'Alt Style' element.
3. Can you past other elements into the 'Content' field? (You should be able to).


You may also want to test this PR in Composer via yalc with: https://github.com/guardian/flexible-content/pull/4492
